Sources/Workspace/Workspace.swift

```diff
@@ -41,59 +41,44 @@ public protocol WorkspaceDelegate: AnyObject {
     func repositoryWillUpdate(_ repository: String)
-
     /// The workspace has finished updating this repository.
-    func repositoryDidUpdate(_ repository: String)
+    func repositoryDidUpdate(_ repository: String, duration: DispatchTimeInterval)
 
     /// The workspace has finished updating and all the dependencies are already up-to-date.
     func dependenciesUpToDate()
 
     /// The workspace is about to clone a repository from the local cache to a working directory.
-    // deprecated 04/2021, remove once clients moved over
-    @available(*, deprecated, message: "use willCreateWorkingCopy")
-    func willClone(repository url: String, to path: AbsolutePath)
     func willCreateWorkingCopy(repository url: String, at path: AbsolutePath)
-
     /// The workspace has cloned a repository from the local cache to a working directory. The error indicates whether the operation failed or succeeded.
     // deprecated 04/2021, remove once clients moved over
-    @available(*, deprecated, message: "use didCreateWorkingCopy")
-    func didClone(repository url: String, to path: AbsolutePath, error: Diagnostic?)
     func didCreateWorkingCopy(repository url: String, at path: AbsolutePath, error: Diagnostic?)
 
-    /// The workspace has started cloning this repository  from the local cache to a working directory. This callback is marginally deprecated in favor of the willClone/didClone pair.
-    // deprecated 04/2021, remove once clients moved over
-    @available(*, deprecated, message: "use didCreateWorkingCopy")
-    func cloning(repository url: String)
-
     /// The workspace is about to check out a particular revision of a working directory.
     func willCheckOut(repository url: String, revision: String, at path: AbsolutePath)
-
     /// The workspace has checked out a particular revision of a working directory. The error indicates whether the operation failed or succeeded.
     func didCheckOut(repository url: String, revision: String, at path: AbsolutePath, error: Diagnostic?)
 
-    // deprecated 04/2021, remove once clients moved over
-    /// The workspace is checking out a repository. This callback is marginally deprecated in favor of the willCheckOut/didCheckOut pair.
-    @available(*, deprecated, message: "use didCheckOut")
-    func checkingOut(repository: String, atReference reference: String, to path: AbsolutePath)
-
     /// The workspace is removing this repository because it is no longer needed.
     func removing(repository: String)
 
     /// Called when the resolver is about to be run.
     func willResolveDependencies(reason: WorkspaceResolveReason)
 
+    /// Called when the resolver begins to be compute the version for the repository.
+    func willComputeVersion(package: PackageIdentity, location: String)
+    /// Called when the resolver finished computing the version for the repository.
+    func didComputeVersion(package: PackageIdentity, location: String, version: String, duration: DispatchTimeInterval)
```

Comment:
Adding this callback is particularly great — is it guaranteed that the package is "done" at this point, i.e. checked out to the right version and there is no chance that PubGrub might go back and check out a different version?

Also, I think that the "version" at this point could be anything, right?  A branch name, etc?  I wonder if we should have a more generic name for the parameter.

Reply:
> Adding this callback is particularly great — is it guaranteed that the package is "done" at this point, i.e. checked out to the right version and there is no chance that PubGrub might go back and check out a different version?

afaict it is

Reply:
> Also, I think that the "version" at this point could be anything, right? A branch name, etc? I wonder if we should have a more generic name for the parameter.

I think this is actually only used for actual semantic versions, using branches will trigger a different flow where this will not be called afaict.

Reply:
see https://github.com/apple/swift-package-manager/pull/3427#discussion_r618845179

Reply:
I see.  It makes sense, since these are all resolved messages that are being added.  I think it's fairly useful, at least in a UI, to see what branches individual packages get checked out to — that would be an argument for keeping the output from the checkout callbacks.

Reply:
okay how does this look: 

```
❯ .build/debug/swift-package update
Fetching https://github.com/apple/swift-llbuild.git from cache
Fetched https://github.com/apple/swift-llbuild.git (1.21s)
Fetching https://github.com/apple/swift-tools-support-core.git from cache
Fetched https://github.com/apple/swift-tools-support-core.git (0.83s)
Fetching https://github.com/apple/swift-driver.git from cache
Fetched https://github.com/apple/swift-driver.git (1.03s)
Fetching https://github.com/apple/swift-argument-parser.git from cache
Fetching https://github.com/apple/swift-crypto.git from cache
Fetching https://github.com/jpsim/Yams.git from cache
Fetched https://github.com/apple/swift-crypto.git (0.88s)
Fetched https://github.com/jpsim/Yams.git (0.93s)
Fetched https://github.com/apple/swift-argument-parser.git (0.97s)
Computing version for https://github.com/apple/swift-argument-parser.git
Computed https://github.com/apple/swift-argument-parser.git at 0.3.2 (0.04s)
Computing version for https://github.com/apple/swift-crypto.git
Computed https://github.com/apple/swift-crypto.git at 1.1.6 (0.03s)
Computing version for https://github.com/jpsim/Yams.git
Computed https://github.com/jpsim/Yams.git at 4.0.6 (0.04s)
Creating working copy for https://github.com/apple/swift-llbuild.git
Resolved https://github.com/apple/swift-llbuild.git at main
Creating working copy for https://github.com/jpsim/Yams.git
Resolved https://github.com/jpsim/Yams.git at 4.0.6
Creating working copy for https://github.com/apple/swift-driver.git
Resolved https://github.com/apple/swift-driver.git at main
Creating working copy for https://github.com/apple/swift-argument-parser.git
Resolved https://github.com/apple/swift-argument-parser.git at 0.3.2
Creating working copy for https://github.com/apple/swift-crypto.git
Resolved https://github.com/apple/swift-crypto.git at 1.1.6
Creating working copy for https://github.com/apple/swift-tools-support-core.git
Resolved https://github.com/apple/swift-tools-support-core.git at main
```

Reply:
Looks great, IMO.  Wordier but much more informative.  So we're using "compute" here to mean decide on what version (but not actually do anything in the file system), vs "resolve" to check out, right?  The clients of libSwiftPM should then change to match that.

Reply:
yes, but this is just naming I came up with and happy to change if there are better terminology. 

cc @neonichu @elsh @yim-lee @ktoso @drexin @weissi  for opinions

Reply:
fyi, the next step wit this would be to have an even nicer output instead of the serial list of events we have now

Reply:
Right, absolutely.  My view on this is that the log should contain structured primary data, with display adapters that can then either render as ANSI output, plain text, HTML in a separate file (with CSS incantations to let you collapse or expand sections), etc.  But for the immediate term I think the slightly wordy list is fine.

There's also a distinction between status and log, I think.  From Terminal I want to see colorful progress bars, etc, while still ending up with a file somewhere that I can mine for details afterwards.

Reply:
> yes, but this is just naming I came up with and happy to change if there are better terminology.
> 
> cc @neonichu @elsh @yim-lee @ktoso @drexin @weissi for opinions

I'm wondering if "checking out" wouldn't still be a better term for the actual `git checkout` at the end, since that does match the semantics of the Git operation.

Reply:
Maybe "evaluate"?

```
Evaluating version for https://github.com/apple/swift-argument-parser.git
Evaluated https://github.com/apple/swift-argument-parser.git to 0.3.2 (0.04s)
```

"Compute" is fine though `Computed ... at ...` doesn't sound right. 🤔

Reply:
> I'm wondering if "checking out" wouldn't still be a better term for the actual git checkout at the end, since that does match the semantics of the Git operation.

part of the idea here is to decouple this from git terminology as with SE-0292 some of these will not be git operations at all

Reply:
> I guess I didn't realize until just now that the "working copy resolved" is the matching "did" part of the "creating working copy" message.

its actually not :) they are two distinct actions: creating the working copy, then pointing it to the computed version. In the suggested output about, we are printing on  "willCreateWorkingCopy" and "didCheckout". printing before we creating the working directory is important because sometime its slow so we need to give the user an indication of what is happening. printing when checkout is complete is to let the user know which revision (version, branch) we landed on eventually and is actually not super useful other than when using branch dependencies.